testing/unittests/jthreadtests.cpp

```diff
@@ -0,0 +1,591 @@
+        pool.setown(createThreadPool("TestPool", factory, true, nullptr, 2));
+
+        // The two threads should start immediately
+        pool->start(&lifespan5seconds, "Thread1", INFINITE);
+        pool->start(&lifespan5seconds, "Thread2", INFINITE);
+        // The third thread should start after one of the above completes
+        std::thread asyncStart([this]()
+                               { this->pool->start(&lifespan1second, "Thread3", INFINITE); });
+        asyncStart.detach();
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        CPPUNIT_ASSERT_EQUAL(2U, threadStartedCount.load());
+
+        std::this_thread::sleep_for(std::chrono::seconds(5));
+        CPPUNIT_ASSERT_EQUAL(3U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(2U, threadCompletedCount.load());
+        CPPUNIT_ASSERT_EQUAL(1U, pool->runningCount());
+
+        // All threads should complete including Thread3
+        pool->joinAll(true);
+        CPPUNIT_ASSERT_EQUAL(3U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(3U, threadCompletedCount.load());
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+        CPPUNIT_ASSERT(pool->running());
+    }
+
+    void testThrottledThreadPool()
+    {
+        // Test: Expected throttling delays are introduced when pool is at capacity
+        pool.setown(createThreadPool("ThrottledTestPool", factory, true, nullptr, 3, 500));
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        unsigned lifespan2seconds = 2;
+        pool->start(&lifespan2seconds, "Thread1");
+        pool->start(&lifespan2seconds, "Thread2");
+        pool->start(&lifespan2seconds, "Thread3");
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+        auto afterQuickStarts = std::chrono::high_resolution_clock::now();
+        auto quickStartDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterQuickStarts - startTime);
+
+        // These should start immediately without throttling
+        CPPUNIT_ASSERT(quickStartDuration.count() < 100);
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        // When pool is at capacity, startNoBlock should throw exception immediately
+        try
+        {
+            pool->startNoBlock(&lifespan2seconds);
+            CPPUNIT_FAIL("startNoBlock should have thrown an exception when pool is at capacity");
+        }
+        catch (IException *e)
+        {
+            StringBuffer msg;
+            e->errorMessage(msg);
+            CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPool", msg.str());
+            e->Release();
+        }
+
+        auto throttleStartTime = std::chrono::high_resolution_clock::now();
+
+        // This should block for approximately 100ms (start delay) then throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4", 100); // timeout after 100ms
```

Comment:
arguably, if there are 2 test functions, 1 for tightly bound, 1 for throttled testing, then this test belongs more in the tightly bound realm